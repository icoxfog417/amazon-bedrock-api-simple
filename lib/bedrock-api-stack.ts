import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';


export class BedrockAPIStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create Lambda function
    const bedrockFunction = new cdk.aws_lambda_nodejs.NodejsFunction(this, 'BedrockFunction', {
      runtime: cdk.aws_lambda.Runtime.NODEJS_LATEST,
      handler: 'handler',
      entry: 'lambda/bedrock-function.ts',
      timeout: cdk.Duration.seconds(30),
      memorySize: 256
    });

    // Grant Bedrock invoke permissions to Lambda
    bedrockFunction.addToRolePolicy(
      new cdk.aws_iam.PolicyStatement({
        effect: cdk.aws_iam.Effect.ALLOW,
        actions: [
          'bedrock:InvokeModel',
          'bedrock:InvokeModelWithResponseStream'
        ],
        resources: [
          `arn:aws:bedrock:*::foundation-model/*`,
          `arn:aws:bedrock:${this.region}:${this.account}:inference-profile/*`
        ],
      })
    );

    // Create API Gateway
    const api = new cdk.aws_apigateway.RestApi(this, 'BedrockApi', {
      restApiName: 'Bedrock API',
      description: 'API Gateway endpoint for Bedrock integration',
    });

    // Create API resource and method
    const predictResource = api.root.addResource('predict');
    predictResource.addMethod('POST', 
      new cdk.aws_apigateway.LambdaIntegration(bedrockFunction), {
        methodResponses: [{
          statusCode: '200',
          responseModels: {
            'application/json': cdk.aws_apigateway.Model.EMPTY_MODEL,
          },
        }],
      }
    );

    new cdk.CfnOutput(this, 'ApiEndpoint', {
      value: api.url,
      description: 'Bedrock API endpoint URL',
    });

  }
}
