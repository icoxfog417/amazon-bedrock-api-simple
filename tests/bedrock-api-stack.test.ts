import * as cdk from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { BedrockAPIStack } from '../lib/bedrock-api-stack';

describe('BedrockAPIStack Tests', () => {
  let app: cdk.App;
  let stack: BedrockAPIStack;
  let template: Template;

  beforeEach(() => {
    // GIVEN
    app = new cdk.App();
    // WHEN
    stack = new BedrockAPIStack(app, 'TestBedrockAPIStack');
    // THEN
    template = Template.fromStack(stack);
  });
  
  test('Stack creates all expected resources', () => {
    // This test ensures we have the correct number of each resource type
    template.resourceCountIs('AWS::Lambda::Function', 1);
    template.resourceCountIs('AWS::ApiGateway::RestApi', 1);
    template.resourceCountIs('AWS::ApiGateway::Resource', 1);
    template.resourceCountIs('AWS::ApiGateway::Method', 1);
    template.resourceCountIs('AWS::ApiGateway::Deployment', 1);
    template.resourceCountIs('AWS::ApiGateway::Stage', 1);
  });

  test('Lambda Function Created with correct properties', () => {
    // Verify Lambda function exists with correct configuration
    template.hasResourceProperties('AWS::Lambda::Function', {
      Handler: 'index.handler', // The handler name from the implementation
      Runtime: 'nodejs18.x', // NODEJS_LATEST resolves to nodejs20.x
      Timeout: 30,
      MemorySize: 256
    });
  });

  test('Lambda Function has correct IAM permissions for Bedrock', () => {
    // Verify IAM policy is created with proper Bedrock permissions
    template.hasResourceProperties('AWS::IAM::Policy', {
      PolicyDocument: {
        Statement: [
          {
            Action: [
              "bedrock:InvokeModel",
              "bedrock:InvokeModelWithResponseStream"
            ],
            Effect: "Allow",
            Resource: [
              "arn:aws:bedrock:*::foundation-model/*",
              {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:bedrock:",
                    {
                      "Ref": "AWS::Region"
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId"
                    },
                    ":inference-profile/*"
                  ]
                ]
              }
            ]
          }
        ]
      }
    });
  });

  test('API Gateway REST API Created', () => {
    // Verify API Gateway exists with correct configuration
    template.hasResourceProperties('AWS::ApiGateway::RestApi', {
      Name: 'Bedrock API',
      Description: 'API Gateway endpoint for Bedrock integration'
    });
  });

  test('API Gateway resource and POST method created', () => {    
    // Verify API Gateway method is created with correct HTTP method and path
    template.hasResourceProperties('AWS::ApiGateway::Method', {
      HttpMethod: 'POST',
      AuthorizationType: 'NONE',
      ResourceId: {
        "Ref": Match.anyValue()
      },
      RestApiId: {
        "Ref": Match.anyValue()
      }
    });
  });

  test('API Gateway integrated with Lambda function', () => {
    // Verify integration exists between API Gateway and Lambda
    template.hasResourceProperties('AWS::ApiGateway::Method', {
      Integration: {
        Type: 'AWS_PROXY',
        IntegrationHttpMethod: 'POST'
      }
    });
  });

  test('API Endpoint output is created', () => {
    // Verify that API endpoint URL is exposed as stack output
    template.hasOutput('ApiEndpoint', {
      Description: 'Bedrock API endpoint URL'
    });
  });
  
  // Add a snapshot test as an additional safety net
  test('Stack matches snapshot', () => {
    // This helps catch any unintended changes to the infrastructure
    expect(JSON.stringify(template.toJSON())).toMatchSnapshot();
  });
});
