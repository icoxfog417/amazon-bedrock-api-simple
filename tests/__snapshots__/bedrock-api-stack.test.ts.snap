// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BedrockAPIStack Tests Stack matches snapshot 1`] = `"{"Resources":{"BedrockFunctionServiceRole02A94671":{"Type":"AWS::IAM::Role","Properties":{"AssumeRolePolicyDocument":{"Statement":[{"Action":"sts:AssumeRole","Effect":"Allow","Principal":{"Service":"lambda.amazonaws.com"}}],"Version":"2012-10-17"},"ManagedPolicyArns":[{"Fn::Join":["",["arn:",{"Ref":"AWS::Partition"},":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"]]}]}},"BedrockFunctionServiceRoleDefaultPolicyBD4AB7A9":{"Type":"AWS::IAM::Policy","Properties":{"PolicyDocument":{"Statement":[{"Action":["bedrock:InvokeModel","bedrock:InvokeModelWithResponseStream"],"Effect":"Allow","Resource":["arn:aws:bedrock:*::foundation-model/*",{"Fn::Join":["",["arn:aws:bedrock:",{"Ref":"AWS::Region"},":",{"Ref":"AWS::AccountId"},":inference-profile/*"]]}]}],"Version":"2012-10-17"},"PolicyName":"BedrockFunctionServiceRoleDefaultPolicyBD4AB7A9","Roles":[{"Ref":"BedrockFunctionServiceRole02A94671"}]}},"BedrockFunctionADEDD58A":{"Type":"AWS::Lambda::Function","Properties":{"Code":{"S3Bucket":{"Fn::Sub":"cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}"},"S3Key":"4530ce3e78a570eee686471744c6e9d0fc6cf0b46e2f2c8980dd1553bb1076ef.zip"},"Handler":"index.handler","MemorySize":256,"Role":{"Fn::GetAtt":["BedrockFunctionServiceRole02A94671","Arn"]},"Runtime":"nodejs18.x","Timeout":30},"DependsOn":["BedrockFunctionServiceRoleDefaultPolicyBD4AB7A9","BedrockFunctionServiceRole02A94671"]},"BedrockApi7EC59F6D":{"Type":"AWS::ApiGateway::RestApi","Properties":{"Description":"API Gateway endpoint for Bedrock integration","Name":"Bedrock API"}},"BedrockApiCloudWatchRole634E5E8A":{"Type":"AWS::IAM::Role","Properties":{"AssumeRolePolicyDocument":{"Statement":[{"Action":"sts:AssumeRole","Effect":"Allow","Principal":{"Service":"apigateway.amazonaws.com"}}],"Version":"2012-10-17"},"ManagedPolicyArns":[{"Fn::Join":["",["arn:",{"Ref":"AWS::Partition"},":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs"]]}]},"UpdateReplacePolicy":"Retain","DeletionPolicy":"Retain"},"BedrockApiAccount7733BA59":{"Type":"AWS::ApiGateway::Account","Properties":{"CloudWatchRoleArn":{"Fn::GetAtt":["BedrockApiCloudWatchRole634E5E8A","Arn"]}},"DependsOn":["BedrockApi7EC59F6D"],"UpdateReplacePolicy":"Retain","DeletionPolicy":"Retain"},"BedrockApiDeployment7919E4191155ed1aa1418871699b041a166a6fa5":{"Type":"AWS::ApiGateway::Deployment","Properties":{"Description":"API Gateway endpoint for Bedrock integration","RestApiId":{"Ref":"BedrockApi7EC59F6D"}},"DependsOn":["BedrockApipredictPOST35CF6E85","BedrockApipredict93D3B161"]},"BedrockApiDeploymentStageprodE6A4312F":{"Type":"AWS::ApiGateway::Stage","Properties":{"DeploymentId":{"Ref":"BedrockApiDeployment7919E4191155ed1aa1418871699b041a166a6fa5"},"RestApiId":{"Ref":"BedrockApi7EC59F6D"},"StageName":"prod"},"DependsOn":["BedrockApiAccount7733BA59"]},"BedrockApipredict93D3B161":{"Type":"AWS::ApiGateway::Resource","Properties":{"ParentId":{"Fn::GetAtt":["BedrockApi7EC59F6D","RootResourceId"]},"PathPart":"predict","RestApiId":{"Ref":"BedrockApi7EC59F6D"}}},"BedrockApipredictPOSTApiPermissionTestBedrockAPIStackBedrockApiC0D1F953POSTpredict02C7DF4D":{"Type":"AWS::Lambda::Permission","Properties":{"Action":"lambda:InvokeFunction","FunctionName":{"Fn::GetAtt":["BedrockFunctionADEDD58A","Arn"]},"Principal":"apigateway.amazonaws.com","SourceArn":{"Fn::Join":["",["arn:",{"Ref":"AWS::Partition"},":execute-api:",{"Ref":"AWS::Region"},":",{"Ref":"AWS::AccountId"},":",{"Ref":"BedrockApi7EC59F6D"},"/",{"Ref":"BedrockApiDeploymentStageprodE6A4312F"},"/POST/predict"]]}}},"BedrockApipredictPOSTApiPermissionTestTestBedrockAPIStackBedrockApiC0D1F953POSTpredict657B9485":{"Type":"AWS::Lambda::Permission","Properties":{"Action":"lambda:InvokeFunction","FunctionName":{"Fn::GetAtt":["BedrockFunctionADEDD58A","Arn"]},"Principal":"apigateway.amazonaws.com","SourceArn":{"Fn::Join":["",["arn:",{"Ref":"AWS::Partition"},":execute-api:",{"Ref":"AWS::Region"},":",{"Ref":"AWS::AccountId"},":",{"Ref":"BedrockApi7EC59F6D"},"/test-invoke-stage/POST/predict"]]}}},"BedrockApipredictPOST35CF6E85":{"Type":"AWS::ApiGateway::Method","Properties":{"AuthorizationType":"NONE","HttpMethod":"POST","Integration":{"IntegrationHttpMethod":"POST","Type":"AWS_PROXY","Uri":{"Fn::Join":["",["arn:",{"Ref":"AWS::Partition"},":apigateway:",{"Ref":"AWS::Region"},":lambda:path/2015-03-31/functions/",{"Fn::GetAtt":["BedrockFunctionADEDD58A","Arn"]},"/invocations"]]}},"MethodResponses":[{"ResponseModels":{"application/json":"Empty"},"StatusCode":"200"}],"ResourceId":{"Ref":"BedrockApipredict93D3B161"},"RestApiId":{"Ref":"BedrockApi7EC59F6D"}}}},"Outputs":{"BedrockApiEndpoint0AF79013":{"Value":{"Fn::Join":["",["https://",{"Ref":"BedrockApi7EC59F6D"},".execute-api.",{"Ref":"AWS::Region"},".",{"Ref":"AWS::URLSuffix"},"/",{"Ref":"BedrockApiDeploymentStageprodE6A4312F"},"/"]]}},"ApiEndpoint":{"Description":"Bedrock API endpoint URL","Value":{"Fn::Join":["",["https://",{"Ref":"BedrockApi7EC59F6D"},".execute-api.",{"Ref":"AWS::Region"},".",{"Ref":"AWS::URLSuffix"},"/",{"Ref":"BedrockApiDeploymentStageprodE6A4312F"},"/"]]}}},"Parameters":{"BootstrapVersion":{"Type":"AWS::SSM::Parameter::Value<String>","Default":"/cdk-bootstrap/hnb659fds/version","Description":"Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]"}},"Rules":{"CheckBootstrapVersion":{"Assertions":[{"Assert":{"Fn::Not":[{"Fn::Contains":[["1","2","3","4","5"],{"Ref":"BootstrapVersion"}]}]},"AssertDescription":"CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI."}]}}}"`;
